File: mcp-server/internal/auth/oidc.go
package auth

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"os/exec"
	"runtime"
	"strings"
	"time"

	"golang.org/x/oauth2"
)

type OIDCConfig struct {
	Authority string
	ClientID  string
	Scopes    []string
}

type Client struct {
	cfg   OIDCConfig
	store *Store
}

func NewClient(cfg OIDCConfig, store *Store) *Client {
	return &Client{cfg: cfg, store: store}
}

func (c *Client) Login(ctx context.Context) (*Session, error) {
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		return nil, fmt.Errorf("start callback listener: %w", err)
	}
	defer listener.Close()

	port := listener.Addr().(*net.TCPAddr).Port
	redirectURI := fmt.Sprintf("http://127.0.0.1:%d/callback", port)

	verifier, challenge, err := newPKCE()
	if err != nil {
		return nil, err
	}
	state, err := randomURLString(24)
	if err != nil {
		return nil, err
	}

	oauthCfg := c.oauthConfig(redirectURI)
	authURL := oauthCfg.AuthCodeURL(
		state,
		oauth2.SetAuthURLParam("code_challenge", challenge),
		oauth2.SetAuthURLParam("code_challenge_method", "S256"),
	)

	codeCh := make(chan string, 1)
	errCh := make(chan error, 1)
	mux := http.NewServeMux()
	mux.HandleFunc("/callback", func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Query().Get("state") != state {
			http.Error(w, "invalid state", http.StatusBadRequest)
			errCh <- errors.New("invalid oauth state")
			return
		}
		if errMsg := r.URL.Query().Get("error"); errMsg != "" {
			desc := r.URL.Query().Get("error_description")
			http.Error(w, errMsg, http.StatusBadRequest)
			errCh <- fmt.Errorf("oauth error: %s (%s)", errMsg, desc)
			return
		}
		code := r.URL.Query().Get("code")
		if code == "" {
			http.Error(w, "missing code", http.StatusBadRequest)
			errCh <- errors.New("missing authorization code")
			return
		}
		_, _ = io.WriteString(w, "<html><body><h2>YAAIF login complete</h2><p>You can close this window and return to Cursor.</p></body></html>")
		codeCh <- code
	})

	srv := &http.Server{Handler: mux}
	go func() { _ = srv.Serve(listener) }()
	defer func() {
		shutdownCtx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
		defer cancel()
		_ = srv.Shutdown(shutdownCtx)
	}()

	if err := openBrowser(authURL); err != nil {
		return nil, fmt.Errorf("open browser for login (%s): %w", authURL, err)
	}

	var code string
	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case err := <-errCh:
		return nil, err
	case code = <-codeCh:
	case <-time.After(5 * time.Minute):
		return nil, errors.New("login timed out waiting for browser callback")
	}

	tok, err := oauthCfg.Exchange(
		ctx,
		code,
		oauth2.SetAuthURLParam("code_verifier", verifier),
	)
	if err != nil {
		return nil, fmt.Errorf("token exchange: %w", err)
	}

	sess := &Session{
		Tokens: tokenSetFromOAuth(tok),
	}
	if idTok, ok := tok.Extra("id_token").(string); ok {
		sess.Tokens.IDToken = idTok
		subj, email, name := parseIDTokenClaims(idTok)
		sess.Subject = subj
		sess.Email = email
		sess.Name = name
	}
	if err := c.store.Save(sess); err != nil {
		return nil, err
	}
	return sess, nil
}

func (c *Client) Logout() error {
	return c.store.Clear()
}

func (c *Client) Session() (*Session, error) {
	return c.store.Load()
}

func (c *Client) SetTenant(tenantID string) (*Session, error) {
	sess, err := c.store.Load()
	if err != nil {
		return nil, err
	}
	if sess == nil || sess.Tokens.AccessToken == "" {
		return nil, errors.New("not authenticated; call yaaif_login first")
	}
	sess.TenantID = strings.TrimSpace(tenantID)
	if err := c.store.Save(sess); err != nil {
		return nil, err
	}
	return sess, nil
}

func (c *Client) AccessToken(ctx context.Context) (string, *Session, error) {
	sess, err := c.store.Load()
	if err != nil {
		return "", nil, err
	}
	if sess == nil || sess.Tokens.AccessToken == "" {
		return "", nil, errors.New("not authenticated; call yaaif_login first")
	}
	if time.Until(sess.Tokens.Expiry) > 45*time.Second {
		return sess.Tokens.AccessToken, sess, nil
	}
	if sess.Tokens.RefreshToken == "" {
		return "", nil, errors.New("access token expired; call yaaif_login again")
	}
	refreshed, err := c.refresh(ctx, sess)
	if err != nil {
		return "", nil, err
	}
	return refreshed.Tokens.AccessToken, refreshed, nil
}

func (c *Client) refresh(ctx context.Context, sess *Session) (*Session, error) {
	form := url.Values{}
	form.Set("grant_type", "refresh_token")
	form.Set("refresh_token", sess.Tokens.RefreshToken)
	form.Set("client_id", c.cfg.ClientID)

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.tokenURL(), strings.NewReader(form.Encode()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 300 {
		return nil, fmt.Errorf("refresh token failed (%d): %s", resp.StatusCode, string(body))
	}
	var tok oauth2.Token
	var raw map[string]any
	if err := json.Unmarshal(body, &raw); err != nil {
		return nil, err
	}
	access, _ := raw["access_token"].(string)
	refresh, _ := raw["refresh_token"].(string)
	tokenType, _ := raw["token_type"].(string)
	expiresIn, _ := raw["expires_in"].(float64)
	idToken, _ := raw["id_token"].(string)
	tok.AccessToken = access
	tok.RefreshToken = refresh
	tok.TokenType = tokenType
	if expiresIn > 0 {
		tok.Expiry = time.Now().Add(time.Duration(expiresIn) * time.Second)
	}
	if refresh == "" {
		tok.RefreshToken = sess.Tokens.RefreshToken
	}
	sess.Tokens = tokenSetFromOAuth(&tok)
	if idToken != "" {
		sess.Tokens.IDToken = idToken
		subj, email, name := parseIDTokenClaims(idToken)
		if subj != "" {
			sess.Subject = subj
		}
		if email != "" {
			sess.Email = email
		}
		if name != "" {
			sess.Name = name
		}
	}
	if err := c.store.Save(sess); err != nil {
		return nil, err
	}
	return sess, nil
}

func (c *Client) oauthConfig(redirectURI string) *oauth2.Config {
	return &oauth2.Config{
		ClientID:    c.cfg.ClientID,
		RedirectURL: redirectURI,
		Scopes:      c.cfg.Scopes,
		Endpoint: oauth2.Endpoint{
			AuthURL:  c.authority() + "/protocol/openid-connect/auth",
			TokenURL: c.tokenURL(),
		},
	}
}

func (c *Client) authority() string {
	return strings.TrimRight(c.cfg.Authority, "/")
}

func (c *Client) tokenURL() string {
	return c.authority() + "/protocol/openid-connect/token"
}

func tokenSetFromOAuth(tok *oauth2.Token) TokenSet {
	return TokenSet{
		AccessToken:  tok.AccessToken,
		RefreshToken: tok.RefreshToken,
		TokenType:    tok.TokenType,
		Expiry:       tok.Expiry,
	}
}

func newPKCE() (verifier, challenge string, err error) {
	verifier, err = randomURLString(32)
	if err != nil {
		return "", "", err
	}
	sum := sha256.Sum256([]byte(verifier))
	challenge = base64.RawURLEncoding.EncodeToString(sum[:])
	return verifier, challenge, nil
}

func randomURLString(n int) (string, error) {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(b), nil
}

func parseIDTokenClaims(idToken string) (subject, email, name string) {
	parts := strings.Split(idToken, ".")
	if len(parts) < 2 {
		return "", "", ""
	}
	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return "", "", ""
	}
	var claims map[string]any
	if err := json.Unmarshal(payload, &claims); err != nil {
		return "", "", ""
	}
	subject, _ = claims["sub"].(string)
	email, _ = claims["email"].(string)
	name, _ = claims["name"].(string)
	if name == "" {
		name, _ = claims["preferred_username"].(string)
	}
	return subject, email, name
}

func openBrowser(rawURL string) error {
	var cmd *exec.Cmd
	switch runtime.GOOS {
	case "darwin":
		cmd = exec.Command("open", rawURL)
	case "windows":
		cmd = exec.Command("rundll32", "url.dll,FileProtocolHandler", rawURL)
	default:
		cmd = exec.Command("xdg-open", rawURL)
	}
	return cmd.Start()
}

File: packages/mcp/src/auth/oidc.ts
import { createHash, randomBytes } from "node:crypto";
import { createServer } from "node:http";
import type { AddressInfo } from "node:net";
import open from "open";
import { yaaifFetch } from "../client/tls.js";
import type { Config } from "../config.js";
import type { Session, SessionStore, TokenSet } from "./store.js";

export class ReauthRequiredError extends Error {
  readonly code = "reauth_required";
  constructor(message: string) {
    super(message);
    this.name = "ReauthRequiredError";
  }
}

export class IssuerMismatchError extends Error {
  readonly code = "issuer_mismatch";
  constructor(
    message: string,
    readonly sessionAuthority: string,
    readonly configAuthority: string,
  ) {
    super(message);
    this.name = "IssuerMismatchError";
  }
}

function b64url(buf: Buffer): string {
  return buf.toString("base64url");
}

function parseIdToken(idToken: string): { subject?: string; email?: string; name?: string } {
  const parts = idToken.split(".");
  if (parts.length < 2) return {};
  try {
    const payload = JSON.parse(Buffer.from(parts[1], "base64url").toString("utf8")) as Record<string, unknown>;
    return {
      subject: typeof payload.sub === "string" ? payload.sub : undefined,
      email: typeof payload.email === "string" ? payload.email : undefined,
      name:
        typeof payload.name === "string"
          ? payload.name
          : typeof payload.preferred_username === "string"
            ? payload.preferred_username
            : undefined,
    };
  } catch {
    return {};
  }
}

function tokenSetFromJson(raw: Record<string, unknown>, previousRefresh?: string): TokenSet {
  const expiresIn = typeof raw.expires_in === "number" ? raw.expires_in : 3600;
  return {
    access_token: String(raw.access_token ?? ""),
    refresh_token: String(raw.refresh_token ?? previousRefresh ?? ""),
    token_type: String(raw.token_type ?? "Bearer"),
    expiry: new Date(Date.now() + expiresIn * 1000).toISOString(),
    id_token: typeof raw.id_token === "string" ? raw.id_token : undefined,
  };
}

function normalizeAuthority(v: string): string {
  return v.replace(/\/+$/, "").toLowerCase();
}

export class AuthClient {
  constructor(
    private readonly cfg: Config,
    private readonly store: SessionStore,
  ) {}

  assertIssuerMatch(sess: Session | null): void {
    if (!sess?.tokens.access_token || !sess.oidc_authority) return;
    const sessionAuth = normalizeAuthority(sess.oidc_authority);
    const configAuth = normalizeAuthority(this.cfg.oidcAuthority);
    if (sessionAuth && configAuth && sessionAuth !== configAuth) {
      throw new IssuerMismatchError(
        `session was issued for ${sess.oidc_authority} but config points at ${this.cfg.oidcAuthority}; switch profile or login again`,
        sess.oidc_authority,
        this.cfg.oidcAuthority,
      );
    }
  }

  async login(): Promise<{ session: Session; auth_url: string }> {
    await this.store.ensureHome(this.cfg.cursorHome);
    const verifier = b64url(randomBytes(32));
    const challenge = b64url(createHash("sha256").update(verifier).digest());
    const state = b64url(randomBytes(24));

    const server = createServer();
    await new Promise<void>((resolve) => server.listen(0, "127.0.0.1", () => resolve()));
    const port = (server.address() as AddressInfo).port;
    const redirectUri = `http://127.0.0.1:${port}/callback`;

    const authUrl = new URL(`${this.cfg.oidcAuthority}/protocol/openid-connect/auth`);
    authUrl.searchParams.set("client_id", this.cfg.oidcClientId);
    authUrl.searchParams.set("response_type", "code");
    authUrl.searchParams.set("redirect_uri", redirectUri);
    authUrl.searchParams.set("scope", this.cfg.oidcScopes.join(" "));
    authUrl.searchParams.set("state", state);
    authUrl.searchParams.set("code_challenge", challenge);
    authUrl.searchParams.set("code_challenge_method", "S256");
    const authUrlStr = authUrl.toString();

    const code = await new Promise<string>((resolve, reject) => {
      const timer = setTimeout(() => {
        server.close();
        reject(new Error(`login timed out waiting for browser callback. Open this URL manually: ${authUrlStr}`));
      }, 5 * 60 * 1000);

      server.on("request", (req, res) => {
        const url = new URL(req.url ?? "/", redirectUri);
        if (url.pathname !== "/callback") {
          res.writeHead(404).end();
          return;
        }
        if (url.searchParams.get("state") !== state) {
          res.writeHead(400).end("invalid state");
          clearTimeout(timer);
          reject(new Error("invalid oauth state"));
          return;
        }
        const err = url.searchParams.get("error");
        if (err) {
          res.writeHead(400).end(err);
          clearTimeout(timer);
          reject(new Error(`oauth error: ${err}`));
          return;
        }
        const authCode = url.searchParams.get("code");
        if (!authCode) {
          res.writeHead(400).end("missing code");
          clearTimeout(timer);
          reject(new Error("missing authorization code"));
          return;
        }
        res.writeHead(200, { "Content-Type": "text/html; charset=utf-8" });
        res.end(`<!doctype html><html><body style="font-family:system-ui;padding:2rem">
<h2>YAAIF login complete</h2>
<p>Signed in to <code>${this.cfg.oidcAuthority}</code>.</p>
<p>You can close this window and return to Cursor.</p>
</body></html>`);
        clearTimeout(timer);
        resolve(authCode);
        server.close();
      });

      void open(authUrlStr).catch((e) => {
        // Keep listening — user can open the URL manually.
        console.error(`Failed to open browser (${String(e)}). Open this URL: ${authUrlStr}`);
      });
    });

    const body = new URLSearchParams({
      grant_type: "authorization_code",
      code,
      redirect_uri: redirectUri,
      client_id: this.cfg.oidcClientId,
      code_verifier: verifier,
    });
    const tokenRes = await yaaifFetch(`${this.cfg.oidcAuthority}/protocol/openid-connect/token`, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body,
    });
    const raw = (await tokenRes.json()) as Record<string, unknown>;
    if (!tokenRes.ok) {
      throw new Error(`token exchange failed (${tokenRes.status}): ${JSON.stringify(raw)}`);
    }

    const session = await this.persistTokens(raw);
    return { session, auth_url: authUrlStr };
  }

  /**
   * OAuth 2.0 device authorization grant (headless / CI).
   * Requires Keycloak client attribute oauth2.device.authorization.grant.enabled=true.
   */
  async deviceLogin(opts: { timeout_ms?: number } = {}): Promise<{
    session: Session;
    verification_uri: string;
    user_code: string;
  }> {
    await this.store.ensureHome(this.cfg.cursorHome);
    const deviceEndpoint = `${this.cfg.oidcAuthority}/protocol/openid-connect/auth/device`;
    const startBody = new URLSearchParams({
      client_id: this.cfg.oidcClientId,
      scope: this.cfg.oidcScopes.join(" "),
    });
    const startRes = await yaaifFetch(deviceEndpoint, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: startBody,
    });
    const startRaw = (await startRes.json()) as Record<string, unknown>;
    if (!startRes.ok) {
      throw new Error(
        `device auth start failed (${startRes.status}): ${JSON.stringify(startRaw)}. ` +
          "Enable oauth2.device.authorization.grant.enabled on the yaaif-cursor Keycloak client.",
      );
    }
    const deviceCode = String(startRaw.device_code ?? "");
    const userCode = String(startRaw.user_code ?? "");
    const verificationUri = String(
      startRaw.verification_uri_complete ?? startRaw.verification_uri ?? "",
    );
    const intervalSec = typeof startRaw.interval === "number" ? startRaw.interval : 5;
    const timeoutMs = opts.timeout_ms ?? 5 * 60 * 1000;
    const deadline = Date.now() + timeoutMs;

    if (verificationUri) {
      void open(verificationUri).catch(() => {
        console.error(`Complete device login at: ${verificationUri} code=${userCode}`);
      });
    }

    while (Date.now() < deadline) {
      await new Promise((r) => setTimeout(r, Math.max(intervalSec, 2) * 1000));
      const pollBody = new URLSearchParams({
        grant_type: "urn:ietf:params:oauth:grant-type:device_code",
        device_code: deviceCode,
        client_id: this.cfg.oidcClientId,
      });
      const tokenRes = await yaaifFetch(`${this.cfg.oidcAuthority}/protocol/openid-connect/token`, {
        method: "POST",
        headers: { "Content-Type": "application/x-www-form-urlencoded" },
        body: pollBody,
      });
      const raw = (await tokenRes.json()) as Record<string, unknown>;
      if (tokenRes.ok && raw.access_token) {
        const session = await this.persistTokens(raw);
        return { session, verification_uri: verificationUri, user_code: userCode };
      }
      const err = String(raw.error ?? "");
      if (err === "authorization_pending" || err === "slow_down") continue;
      throw new Error(`device auth poll failed: ${JSON.stringify(raw)}`);
    }
    throw new Error(
      `device login timed out. Open ${verificationUri} and enter code ${userCode}`,
    );
  }

  private async persistTokens(raw: Record<string, unknown>): Promise<Session> {
    const tokens = tokenSetFromJson(raw);
    const claims = tokens.id_token ? parseIdToken(tokens.id_token) : {};
    const session: Session = {
      tokens,
      subject: claims.subject,
      email: claims.email,
      name: claims.name,
      tenant_id: this.cfg.defaultTenantId || undefined,
      profile_id: this.cfg.activeProfileId || undefined,
      oidc_authority: this.cfg.oidcAuthority,
    };
    await this.store.save(session);
    return session;
  }

  async logout(opts: { endSession?: boolean } = {}): Promise<{ logged_out: boolean; end_session_url?: string }> {
    const sess = await this.store.load();
    let end_session_url: string | undefined;
    if (opts.endSession && sess?.tokens.id_token && sess.oidc_authority) {
      const u = new URL(`${sess.oidc_authority}/protocol/openid-connect/logout`);
      u.searchParams.set("id_token_hint", sess.tokens.id_token);
      end_session_url = u.toString();
      void open(end_session_url).catch(() => { /* optional */ });
    }
    await this.store.clear();
    return { logged_out: true, end_session_url };
  }

  async session(): Promise<Session | null> {
    return this.store.load();
  }

  async setTenant(tenantId: string, tenantName?: string): Promise<Session> {
    const sess = await this.store.load();
    if (!sess?.tokens.access_token) throw new ReauthRequiredError("not authenticated; call yaaif_login first");
    this.assertIssuerMatch(sess);
    sess.tenant_id = tenantId.trim();
    if (tenantName) sess.tenant_name = tenantName.trim();
    await this.store.save(sess);
    return sess;
  }

  async accessToken(): Promise<{ token: string; session: Session }> {
    let sess = await this.store.load();
    if (!sess?.tokens.access_token) throw new ReauthRequiredError("not authenticated; call yaaif_login first");
    this.assertIssuerMatch(sess);
    const expiry = Date.parse(sess.tokens.expiry);
    if (Number.isFinite(expiry) && expiry - Date.now() > 45_000) {
      return { token: sess.tokens.access_token, session: sess };
    }
    if (!sess.tokens.refresh_token) {
      await this.store.clear();
      throw new ReauthRequiredError("access token expired and no refresh token; call yaaif_login again");
    }
    try {
      sess = await this.refresh(sess);
    } catch (e) {
      await this.store.clear();
      throw new ReauthRequiredError(`token refresh failed; call yaaif_login again (${String(e)})`);
    }
    return { token: sess.tokens.access_token, session: sess };
  }

  private async refresh(sess: Session): Promise<Session> {
    const authority = sess.oidc_authority || this.cfg.oidcAuthority;
    const body = new URLSearchParams({
      grant_type: "refresh_token",
      refresh_token: sess.tokens.refresh_token ?? "",
      client_id: this.cfg.oidcClientId,
    });
    const tokenRes = await yaaifFetch(`${authority}/protocol/openid-connect/token`, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body,
    });
    const raw = (await tokenRes.json()) as Record<string, unknown>;
    if (!tokenRes.ok) {
      throw new Error(`refresh token failed (${tokenRes.status}): ${JSON.stringify(raw)}`);
    }
    sess.tokens = tokenSetFromJson(raw, sess.tokens.refresh_token);
    if (sess.tokens.id_token) {
      const claims = parseIdToken(sess.tokens.id_token);
      sess.subject = claims.subject ?? sess.subject;
      sess.email = claims.email ?? sess.email;
      sess.name = claims.name ?? sess.name;
    }
    sess.oidc_authority = authority;
    sess.profile_id = this.cfg.activeProfileId || sess.profile_id;
    await this.store.save(sess);
    return sess;
  }
}
